src/pveclib/vec_int32_ppc.h

```diff
@@ -1419,4 +1300,257 @@ vec_srwi (vui32_t vra, const unsigned int shb)
+ * But for POWER7 and earlier we have to construct word multiplies
+ * from two halfword multiplies (vmuleuh and vmulouh). Then sum the
+ * partial products for the final doubleword results. This is
+ * complicated by the fact that vector add doubleword is not available
+ * for POWER7. So we need to construct the doubleword add from
+ * Vector Add Unsigned Word Modulo (vadduwm) and
+ * Vector Add and Write Carry-Out Unsigned Word (vaddcuw) with
+ * shift double quadword to reposition the low word carry and
+ * a final vadduwm to complete the carry propagation for the
+ * doubleword add.
+ *
+ * \note This implementation is NOT endian sensitive and the function
+ * is stable across BE/LE implementations.
+ *
+ * |processor|Latency|Throughput|
+ * |--------:|:-----:|:---------|
+ * |power8   |   7   | 2/cycle  |
+ * |power9   |   7   | 2/cycle  |
+ *
+ * @param vra 128-bit vector unsigned int.
+ * @param vrb 128-bit vector unsigned int.
+ * @return vector unsigned long product of the even words of a and b.
+ */
+static inline vui64_t
+vec_vmuleuw (vui32_t vra, vui32_t vrb)
+{
+  vui64_t res;
+#ifdef _ARCH_PWR8
+#if defined __GNUC__ && (__GNUC__ > 7)
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+  res = vec_mulo (vra, vrb);
+#else
+  res = vec_mule (vra, vrb);
+#endif
+#else
+  __asm__(
+      "vmuleuw %0,%1,%2;\n"
+      : "=v" (res)
+      : "v" (vra),
+      "v" (vrb)
+      : );
+#endif
+#else
+  const vui32_t zero = {0,0,0,0};
+  const vui32_t ones = {-1,-1,-1,-1};
+  vui32_t wmask01;
+  vui32_t p0, p1, pp10, pp01, resw;
+  vui16_t m0, m1, mt, mth, mtl;
+
+  /* generate {0,-1,0,-1}  mask.  */
+  wmask01 = vec_vmrghw (zero, ones);
+
+  mt = (vui16_t)vrb;
+  mtl = vec_mergeh (mt, mt);
+  mth = vec_mergel (mt, mt);
+
+#ifdef _ARCH_PWR7
+  m0 = (vui16_t)vec_xxpermdi ((vui64_t)mtl, (vui64_t)mth, 0);
+#else
+  {
+    vui32_t temp;
+    temp = vec_sld ((vui32_t) mtl, (vui32_t) mth, 8);
+    m0 = (vui16_t) vec_sld (temp, (vui32_t) mth, 8);
+  }
+#endif
+
+  resw = vec_sld (vra, vra, 12);
+  resw = vec_sel (vra, resw, wmask01);
+  m1 = (vui16_t)resw;
+
+  p0 = vec_vmuleuh (m1, m0);
+  p1 = vec_vmulouh (m1, m0);
+  resw = vec_sel (p0, p1, wmask01);
+  res = (vui64_t)resw;
+
+  pp10 = vec_sld (p1, p1, 12);
+  pp01 = p0;
+  /* pp01 = vec_addudm (pp01, pp10).  */
+  {
+    vui32_t c;
+    vui32_t xmask;
+    xmask = vec_sld (wmask01, wmask01, 2);
+    c    = vec_vaddcuw (pp01, pp10);
+    pp01 = vec_vadduwm (pp01, pp10);
+    c    = vec_sld (c, c, 6);
+    pp01 = vec_sld (pp01, pp01, 2);
+    pp01 = vec_sel (c, pp01, xmask);
+  }
+  /* res = vec_addudm (pp01, res).  */
+  {
+    vui32_t c, r;
+    c = vec_vaddcuw (pp01, (vui32_t)res);
+    r = vec_vadduwm (pp01, (vui32_t)res);
+    c = vec_sld (c, zero, 4);
+    res = (vui64_t)vec_vadduwm (r, c);
+  }
+#endif
+  return (res);
+}
+
+/** \brief Vector Multiply Odd Unsigned Words.
+ *
+ * Multiple the odd words of two vector unsigned int values and return
```

Comment:
s/Multiple/Multiply/
Also, same comments with respect to bit size of types.
So, these are now global comments.

Reply:
Ok, see above.